In [1]:
import pandas as pd
import numpy as np 
import bs4 as bs
import urllib.request
!pip install requests
import requests

In [2]:
url='https://en.wikipedia.org/wiki/Municipalities_of_Milan'

response=requests.get(url)

html = response.text

page = bs.BeautifulSoup(html,'html.parser')

print(page)


<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>Municipalities of Milan - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"b401fc65-f296-48bc-acb3-0c662400233f","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"Municipalities_of_Milan","wgTitle":"Municipalities of Milan","wgCurRevisionId":959118297,"wgRevisionId":959118297,"wgArticleId":29822831,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with Italian-language sources (it)","Webarchive template wayback links","Articles containing Italian-language text","Commons ca

In [3]:
Milan = page.find('table',{'class':'wikitable sortable'})
Milan

<table class="wikitable sortable">
<tbody><tr>
<th>Borough</th>
<th class="unsortable">Name</th>
<th>Area<br/>(km<sup>2</sup>)</th>
<th>Population<br/>(2014)</th>
<th>Population density<br/>(inhabitants/km<sup>2</sup>)</th>
<th>Quartieri (districts)
</th></tr>
<tr>
<td><i><a href="/wiki/Zone_1_of_Milan" title="Zone 1 of Milan">1</a></i></td>
<td><a class="new" href="/w/index.php?title=Centro_storico_(district_of_Milan)&amp;action=edit&amp;redlink=1" title="Centro storico (district of Milan) (page does not exist)">Centro storico</a></td>
<td>9.67</td>
<td>96,315</td>
<td>11,074</td>
<td><a href="/wiki/Brera_(district_of_Milan)" title="Brera (district of Milan)">Brera</a>, <a class="new" href="/w/index.php?title=Centro_Storico_(district_of_Milan)&amp;action=edit&amp;redlink=1" title="Centro Storico (district of Milan) (page does not exist)">Centro Storico</a>, <a href="/wiki/Conca_del_Naviglio" title="Conca del Naviglio">Conca del Naviglio</a>, <a class="new" href="/w/index.php?title=Gua

In [4]:
rows=list()
for row in Milan.findAll("tr"):
   rows.append(row)
rows

[<tr>
 <th>Borough</th>
 <th class="unsortable">Name</th>
 <th>Area<br/>(km<sup>2</sup>)</th>
 <th>Population<br/>(2014)</th>
 <th>Population density<br/>(inhabitants/km<sup>2</sup>)</th>
 <th>Quartieri (districts)
 </th></tr>, <tr>
 <td><i><a href="/wiki/Zone_1_of_Milan" title="Zone 1 of Milan">1</a></i></td>
 <td><a class="new" href="/w/index.php?title=Centro_storico_(district_of_Milan)&amp;action=edit&amp;redlink=1" title="Centro storico (district of Milan) (page does not exist)">Centro storico</a></td>
 <td>9.67</td>
 <td>96,315</td>
 <td>11,074</td>
 <td><a href="/wiki/Brera_(district_of_Milan)" title="Brera (district of Milan)">Brera</a>, <a class="new" href="/w/index.php?title=Centro_Storico_(district_of_Milan)&amp;action=edit&amp;redlink=1" title="Centro Storico (district of Milan) (page does not exist)">Centro Storico</a>, <a href="/wiki/Conca_del_Naviglio" title="Conca del Naviglio">Conca del Naviglio</a>, <a class="new" href="/w/index.php?title=Guastalla_(district_of_Milan)&

In [5]:
milan_df = []
for tr in rows:
    td = tr.find_all('td')
    row = [tr.text for tr in td]
    milan_df.append(row)
milan_df=pd.DataFrame(milan_df, columns=["Borough No.", "Borough Name", 'Area_km2','Population','Density population_inhabitans_km2',"Neighbourhood"])
milan_df

Borough No.                                       Borough Name Area_km2  \
0         None                                               None     None   
1            1                                     Centro storico     9.67   
2            2  Stazione Centrale, Gorla, Turro, Greco, Cresce...    12.58   
3            3               Città Studi, Lambrate, Porta Venezia    14.23   
4            4                          Porta Vittoria, Forlanini    20.95   
5            5                Vigentino, Chiaravalle, Gratosoglio    29.87   
6            6                                Barona, Lorenteggio    18.28   
7            7                        Baggio, De Angeli, San Siro    31.34   
8            8                 Fiera, Gallaratese, Quarto Oggiaro    23.72   
9            9                          Porta Garibaldi, Niguarda    21.12   
10        None                                               None     None   

   Population Density population_inhabitans_km2  \
0        None                              None   
1      96,315                            11,074   
2     153.109                            13,031   
3     141,229                            10,785   
4     156.369                             8,069   
5     123,779                             4,487   
6     149,000                             8,998   
7     170,814                             6,093   
8     181,669                             8,326   
9     181,598                             9,204   
10       None                              None   

                                        Neighbourhood  
0                                                None  
1   Brera, Centro Storico, Conca del Naviglio, Gua...  
2   Adriano, Crescenzago, Gorla, Greco, Loreto, Ma...  
3   Casoretto, Cimiano, Città Studi, Dosso, Lambra...  
4   Acquabella, Calvairate, Castagnedo, Cavriano, ...  
5   Basmetto, Cantalupa, Case Nuove, Chiaravalle, ...  
6   Arzaga, Barona, Boffalora, Cascina Bianca, Con...  
7   Assiano, Baggio, Figino, Fopponino, Forze Arma...  
8   Boldinasco, Bullona, Cagnola, Campo dei Fiori,...  
9   Affori, Bicocca, Bovisa, Bovisasca, Bruzzano, ...  
10                                               None

In [6]:
milan_df.dropna(subset=["Borough Name"], axis=0, inplace=True)
milan_df.head()

Borough No.                                       Borough Name Area_km2  \
1           1                                     Centro storico     9.67   
2           2  Stazione Centrale, Gorla, Turro, Greco, Cresce...    12.58   
3           3               Città Studi, Lambrate, Porta Venezia    14.23   
4           4                          Porta Vittoria, Forlanini    20.95   
5           5                Vigentino, Chiaravalle, Gratosoglio    29.87   

  Population Density population_inhabitans_km2  \
1     96,315                            11,074   
2    153.109                            13,031   
3    141,229                            10,785   
4    156.369                             8,069   
5    123,779                             4,487   

                                       Neighbourhood  
1  Brera, Centro Storico, Conca del Naviglio, Gua...  
2  Adriano, Crescenzago, Gorla, Greco, Loreto, Ma...  
3  Casoretto, Cimiano, Città Studi, Dosso, Lambra...  
4  Acquabella, Calvairate, Castagnedo, Cavriano, ...  
5  Basmetto, Cantalupa, Case Nuove, Chiaravalle, ...

Split neighbourhood into multiple rows to differentiate them aming each other

In [7]:
lst_col = 'Neighbourhood'

In [8]:
x = milan_df.assign(**{lst_col:milan_df[lst_col].str.split(',')})

In [9]:
x1=pd.DataFrame({col:np.repeat(x[col].values, x[lst_col].str.len()) for col in x.columns.difference([lst_col])}).assign(**{lst_col:np.concatenate(x[lst_col].values)})[x.columns.tolist()]

In [10]:
x1.head()

Borough No.    Borough Name Area_km2 Population  \
0           1  Centro storico     9.67     96,315   
1           1  Centro storico     9.67     96,315   
2           1  Centro storico     9.67     96,315   
3           1  Centro storico     9.67     96,315   
4           1  Centro storico     9.67     96,315   

  Density population_inhabitans_km2        Neighbourhood  
0                            11,074                Brera  
1                            11,074       Centro Storico  
2                            11,074   Conca del Naviglio  
3                            11,074            Guastalla  
4                            11,074       Porta Sempione

In [11]:
x1.rename(columns={'newName1': 'Neighbourhood'}, inplace=True)

In [12]:
x1.rename(columns={'Neighbourhood': 'Neighbourhoods'}, inplace=True)
x1.head()

Borough No.    Borough Name Area_km2 Population  \
0           1  Centro storico     9.67     96,315   
1           1  Centro storico     9.67     96,315   
2           1  Centro storico     9.67     96,315   
3           1  Centro storico     9.67     96,315   
4           1  Centro storico     9.67     96,315   

  Density population_inhabitans_km2       Neighbourhoods  
0                            11,074                Brera  
1                            11,074       Centro Storico  
2                            11,074   Conca del Naviglio  
3                            11,074            Guastalla  
4                            11,074       Porta Sempione

In [13]:
!pip install geocoder

     |████████████████████████████████| 102kB 7.7MB/s ta 0:00:011


In [14]:
import geocoder

In [15]:
# The code was removed by Watson Studio for sharing.

Neighbourhood  Postal Code
0               Brera        20121
1  Conca del Naviglio        20123
2           Guastalla        20151
3      Porta Sempione        20145
4    Porta Tenaglia\n        20121

In [16]:
df_data_0.rename(columns={'Neighbourhood': 'Neighbourhoods'}, inplace=True)
df_data_0.head()

Neighbourhoods  Postal Code
0               Brera        20121
1  Conca del Naviglio        20123
2           Guastalla        20151
3      Porta Sempione        20145
4    Porta Tenaglia\n        20121

In [17]:
df_data_0 ['Neighbourhoods'].astype(str)
df_data_0['Neighbourhoods'] = df_data_0['Neighbourhoods'].str.replace('\n','')
df_data_0.head()

Neighbourhoods  Postal Code
0               Brera        20121
1  Conca del Naviglio        20123
2           Guastalla        20151
3      Porta Sempione        20145
4    Porta Tenaglia\n        20121

In [18]:
x1['Neighbourhoods'] = x1['Neighbourhoods'].astype(str)
x1['Neighbourhoods'].astype(str)
x1.dtypes

Borough No.                          object
Borough Name                         object
Area_km2                             object
Population                           object
Density population_inhabitans_km2    object
Neighbourhoods                       object
dtype: object

In [19]:
df_data_0.dtypes

Neighbourhoods    object
Postal Code        int64
dtype: object

In [20]:
x2 = x1[['Neighbourhoods']]
x2.head()

Neighbourhoods
0                Brera
1       Centro Storico
2   Conca del Naviglio
3            Guastalla
4       Porta Sempione

In [21]:
x2.dtypes

Neighbourhoods    object
dtype: object

In [22]:
Combined_trial= df_data_0.join(x2.set_index('Neighbourhoods'), on='Neighbourhoods')
Combined_trial

Neighbourhoods  Postal Code
0                         Brera        20121
1            Conca del Naviglio        20123
2                     Guastalla        20151
3                Porta Sempione        20145
4              Porta Tenaglia\n        20121
5                       Adriano        20128
6                   Crescenzago        20134
7                         Gorla        20126
8                         Greco        20125
9                        Loreto        20131
10                   Maggiolina        20125
11                     Mandello        20123
12                    Mirabello        20121
13                 Ponte Seveso        20125
14                  Porta Nuova        20125
15                     Precotto        20128
16            Stazione Centrale        20125
17                        Turro        20127
18  Villaggio dei Giornalisti\n        20129
19                    Casoretto        20131
20                      Cimiano        20132
21                  Città Studi        20077
22                        Dosso        20132
23                     Lambrate        20131
24                       Ortica        20134
25               Porta Monforte        20122
26                Porta Venezia        20143
27             Quartiere Feltre        20132
28                    Rottole\n        20132
29                   Lampugnano        20151
30                      Musocco        20151
31                  Porta Volta        20151
32                     Portello        20151
33               Quarto Oggiaro        20151
34                          QT8        20151
35                      Roserio        20151
36                 San Leonardo        20151
37                       Trenno        20151
38                     Varesina        20151
39                       Vialba        20151
40               Villapizzone\n        20151
41                       Affori        20161
42                      Bicocca        20126
43                       Bovisa        20077
44                    Bovisasca        20157
45                     Bruzzano        20161
46                   Ca' Granda        20162
47           Centro Direzionale        20090
48                     Comasina        20161
49                      Dergano        20154
50                 Fulvio Testi        20124
51                        Isola        20077
52                   La Fontana        20122
53                   Montalbino        20159
54                     Niguarda        20162
55              Porta Garibaldi        20154
56                  Porta Nuova        20125
57              Prato Centenaro        20125
58                    Segnano\n        20060

In [23]:
def get_latilong(postal_code):
    lati_long_coords = None
    while(lati_long_coords is None):
        g = geocoder.arcgis('{}, Milan, Italy'.format(postal_code))
        lati_long_coords = g.latlng
    return lati_long_coords
    
get_latilong('20161')

[45.520290000000045, 9.16698857700004]

In [24]:
# Retrieving Postal Code Co-ordinates
postal_codes = Combined_trial ['Postal Code']    
coords = [ get_latilong(postal_code) for postal_code in postal_codes.tolist() ]

In [25]:
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])
Combined_trial['Latitude'] = df_coords['Latitude']
Combined_trial['Longitude'] = df_coords['Longitude']

In [26]:
Combined_trial

Neighbourhoods  Postal Code   Latitude  Longitude
0                         Brera        20121  45.472955   9.187099
1            Conca del Naviglio        20123  45.463065   9.177542
2                     Guastalla        20151  45.494555   9.113667
3                Porta Sempione        20145  45.473887   9.161670
4              Porta Tenaglia\n        20121  45.472955   9.187099
5                       Adriano        20128  45.513235   9.239450
6                   Crescenzago        20134  45.482773   9.254800
7                         Gorla        20126  45.515929   9.215605
8                         Greco        20125  45.499710   9.208823
9                        Loreto        20131  45.487090   9.227062
10                   Maggiolina        20125  45.499710   9.208823
11                     Mandello        20123  45.463065   9.177542
12                    Mirabello        20121  45.472955   9.187099
13                 Ponte Seveso        20125  45.499710   9.208823
14                  Porta Nuova        20125  45.499710   9.208823
15                     Precotto        20128  45.513235   9.239450
16            Stazione Centrale        20125  45.499710   9.208823
17                        Turro        20127  45.497700   9.226106
18  Villaggio dei Giornalisti\n        20129  45.472203   9.213030
19                    Casoretto        20131  45.487090   9.227062
20                      Cimiano        20132  45.503230   9.246742
21                  Città Studi        20077  45.355305   9.320114
22                        Dosso        20132  45.503230   9.246742
23                     Lambrate        20131  45.487090   9.227062
24                       Ortica        20134  45.482773   9.254800
25               Porta Monforte        20122  45.462565   9.198647
26                Porta Venezia        20143  45.444575   9.158881
27             Quartiere Feltre        20132  45.503230   9.246742
28                    Rottole\n        20132  45.503230   9.246742
29                   Lampugnano        20151  45.494555   9.113667
30                      Musocco        20151  45.494555   9.113667
31                  Porta Volta        20151  45.494555   9.113667
32                     Portello        20151  45.494555   9.113667
33               Quarto Oggiaro        20151  45.494555   9.113667
34                          QT8        20151  45.494555   9.113667
35                      Roserio        20151  45.494555   9.113667
36                 San Leonardo        20151  45.494555   9.113667
37                       Trenno        20151  45.494555   9.113667
38                     Varesina        20151  45.494555   9.113667
39                       Vialba        20151  45.494555   9.113667
40               Villapizzone\n        20151  45.494555   9.113667
41                       Affori        20161  45.520290   9.166989
42                      Bicocca        20126  45.515929   9.215605
43                       Bovisa        20077  45.355305   9.320114
44                    Bovisasca        20157  45.516985   9.125239
45                     Bruzzano        20161  45.520290   9.166989
46                   Ca' Granda        20162  45.515339   9.197820
47           Centro Direzionale        20090  45.423010   9.055915
48                     Comasina        20161  45.520290   9.166989
49                      Dergano        20154  45.483920   9.175276
50                 Fulvio Testi        20124  45.485165   9.201855
51                        Isola        20077  45.355305   9.320114
52                   La Fontana        20122  45.462565   9.198647
53                   Montalbino        20159  45.496732   9.189580
54                     Niguarda        20162  45.515339   9.197820
55              Porta Garibaldi        20154  45.483920   9.175276
56                  Porta Nuova        20125  45.499710   9.208823
57              Prato Centenaro        20125  45.499710   9.208823
58                    Segnano\n        20060  45.523685   9.4493

In [27]:
!pip install folium
import folium

     |████████████████████████████████| 102kB 8.3MB/s ta 0:00:011


In [28]:
from geopy.geocoders import Nominatim 

In [29]:

address = 'Milan,Italy'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The Geograpical Co-ordinate of Milan,Italy are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:4: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.


The Geograpical Co-ordinate of Milan,Italy are 45.4668, 9.1905.


In [30]:
# create map of New York using latitude and longitude values
map_Milan = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, neighborhood in zip(Combined_trial['Latitude'], Combined_trial['Longitude'], Combined_trial['Neighbourhoods']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Milan)  
    
map_Milan

In [31]:
# The code was removed by Watson Studio for sharing.

Your credentails:
CLIENT_ID: WH1GAZW1VQLMRGKX1VIFLK200GAHBOOCVKQD4EWQN14UI0GY
CLIENT_SECRET:XJIJROFNUHSVQCSC1PAWGJK5VWCNM2AEIZJVDA4YJFZZBNZQ


In [32]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [33]:
LIMIT = 100
milan_venues = getNearbyVenues(names=Combined_trial['Neighbourhoods'],
                                   latitudes=Combined_trial['Latitude'],
                                   longitudes=Combined_trial['Longitude']
                                  )

Brera
Conca del Naviglio
Guastalla
Porta Sempione
Porta Tenaglia\n
Adriano
Crescenzago
Gorla
Greco
Loreto
Maggiolina
Mandello
Mirabello
Ponte Seveso
Porta Nuova
Precotto
Stazione Centrale
Turro
Villaggio dei Giornalisti\n
Casoretto
Cimiano
Città Studi
Dosso
Lambrate
Ortica
Porta Monforte
Porta Venezia
Quartiere Feltre
Rottole\n
Lampugnano
Musocco
Porta Volta
Portello
Quarto Oggiaro
QT8
Roserio
San Leonardo
Trenno
Varesina
Vialba
Villapizzone\n
Affori
Bicocca
Bovisa
Bovisasca
Bruzzano
Ca' Granda
Centro Direzionale
Comasina
Dergano
Fulvio Testi
Isola
La Fontana
Montalbino
Niguarda
Porta Garibaldi
Porta Nuova
Prato Centenaro
Segnano\n


In [34]:
print(milan_venues.shape)
milan_venues.head(10)

(1352, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0        Brera              45.472955                9.187099   
1        Brera              45.472955                9.187099   
2        Brera              45.472955                9.187099   
3        Brera              45.472955                9.187099   
4        Brera              45.472955                9.187099   
5        Brera              45.472955                9.187099   
6        Brera              45.472955                9.187099   
7        Brera              45.472955                9.187099   
8        Brera              45.472955                9.187099   
9        Brera              45.472955                9.187099   

                     Venue  Venue Latitude  Venue Longitude  \
0      Pinacoteca di Brera       45.471979         9.188128   
1   Di Viole Di Liquirizia       45.471460         9.185336   
2         Palazzo di Brera       45.472019         9.188043   
3                  SUSHI B       45.472153         9.186883   
4            Dimore Studio       45.474770         9.186954   
5  Antica Osteria Stendhal       45.473978         9.187678   
6                Kusmi Tea       45.472053         9.185824   
7           N'Ombra de Vin       45.473452         9.187873   
8         La Prosciutteria       45.474152         9.183449   
9           Palazzo Parigi       45.473378         9.191108   

          Venue Category  
0             Art Museum  
1           Cupcake Shop  
2  College Arts Building  
3    Japanese Restaurant  
4            Art Gallery  
5     Italian Restaurant  
6               Tea Room  
7               Wine Bar  
8         Sandwich Place  
9                  Hotel

In [35]:
milan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                                 
Adriano                                          6                       6   
Affori                                           9                       9   
Bicocca                                         23                      23   
Bovisa                                          12                      12   
Bovisasca                                        1                       1   
Brera                                           49                      49   
Bruzzano                                         9                       9   
Ca' Granda                                       4                       4   
Casoretto                                       53                      53   
Centro Direzionale                               8                       8   
Cimiano                                          7                       7   
Città Studi                                     12                      12   
Comasina                                         9                       9   
Conca del Naviglio                              34                      34   
Crescenzago                                      4                       4   
Dergano                                         45                      45   
Dosso                                            7                       7   
Fulvio Testi                                   100                     100   
Gorla                                           23                      23   
Greco                                           13                      13   
Guastalla                                       16                      16   
Isola                                           12                      12   
La Fontana                                      66                      66   
Lambrate                                        53                      53   
Lampugnano                                      16                      16   
Loreto                                          53                      53   
Maggiolina                                      13                      13   
Mandello                                        34                      34   
Mirabello                                       49                      49   
Montalbino                                      37                      37   
Musocco                                         16                      16   
Niguarda                                         4                       4   
Ortica                                           4                       4   
Ponte Seveso                                    13                      13   
Porta Garibaldi                                 45                      45   
Porta Monforte                                  66                      66   
Porta Nuova                                     26                      26   
Porta Sempione                                  26                      26   
Porta Tenaglia\n                                49                      49   
Porta Venezia                                   38                      38   
Porta Volta                                     16                      16   
Portello                                        16                      16   
Prato Centenaro                                 13                      13   
Precotto                                         6                       6   
QT8                                             16                      16   
Quartiere Feltre                                 7                       7   
Quarto Oggiaro                                  16                      16   
Roserio                                         16                      16   
Rottole\n                                        7                       7   
San Leonardo                                    16 

In [36]:
print('There are {} uniques categories.'.format(len(milan_venues['Venue Category'].unique())))

There are 146 uniques categories.


In [37]:
# one hot encoding
milan_onehot = pd.get_dummies(milan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
milan_onehot['Neighborhood'] = milan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [milan_onehot.columns[-1]] + list(milan_onehot.columns[:-1])
milan_onehot = milan_onehot[fixed_columns]

milan_onehot.head()

Neighborhood  Accessories Store  African Restaurant  Arcade  \
0        Brera                  0                   0       0   
1        Brera                  0                   0       0   
2        Brera                  0                   0       0   
3        Brera                  0                   0       0   
4        Brera                  0                   0       0   

   Argentinian Restaurant  Art Gallery  Art Museum  Asian Restaurant  Bakery  \
0                       0            0           1                 0       0   
1                       0            0           0                 0       0   
2                       0            0           0                 0       0   
3                       0            0           0                 0       0   
4                       0            1           0                 0       0   

   Bar  ...  Tram Station  Trattoria/Osteria  Tunnel  Tuscan Restaurant  \
0    0  ...             0                  0       0                  0   
1    0  ...             0                  0       0                  0   
2    0  ...             0                  0       0                  0   
3    0  ...             0                  0       0                  0   
4    0  ...             0                  0       0                  0   

   Vegetarian / Vegan Restaurant  Vietnamese Restaurant  Watch Shop  Wine Bar  \
0                              0                      0           0         0   
1                              0                      0           0         0   
2                              0                      0           0         0   
3                              0                      0           0         0   
4                              0                      0           0         0   

   Wine Shop  Women's Store  
0          0              0  
1          0              0  
2          0              0  
3          0              0  
4          0              0  

[5 rows x 147 columns]

In [38]:
milan_onehot.shape

(1352, 147)

In [39]:
milan_grouped = milan_onehot.groupby('Neighborhood').mean().reset_index()
milan_grouped

Neighborhood  Accessories Store  African Restaurant  \
0                       Adriano           0.000000            0.000000   
1                        Affori           0.000000            0.000000   
2                       Bicocca           0.000000            0.000000   
3                        Bovisa           0.000000            0.000000   
4                     Bovisasca           0.000000            0.000000   
5                         Brera           0.020408            0.000000   
6                      Bruzzano           0.000000            0.000000   
7                    Ca' Granda           0.000000            0.000000   
8                     Casoretto           0.000000            0.018868   
9            Centro Direzionale           0.000000            0.000000   
10                      Cimiano           0.000000            0.000000   
11                  Città Studi           0.000000            0.000000   
12                     Comasina           0.000000            0.000000   
13           Conca del Naviglio           0.000000            0.000000   
14                  Crescenzago           0.000000            0.000000   
15                      Dergano           0.022222            0.000000   
16                        Dosso           0.000000            0.000000   
17                 Fulvio Testi           0.000000            0.000000   
18                        Gorla           0.000000            0.000000   
19                        Greco           0.000000            0.000000   
20                    Guastalla           0.000000            0.000000   
21                        Isola           0.000000            0.000000   
22                   La Fontana           0.000000            0.000000   
23                     Lambrate           0.000000            0.018868   
24                   Lampugnano           0.000000            0.000000   
25                       Loreto           0.000000            0.018868   
26                   Maggiolina           0.000000            0.000000   
27                     Mandello           0.000000            0.000000   
28                    Mirabello           0.020408            0.000000   
29                   Montalbino           0.000000            0.000000   
30                      Musocco           0.000000            0.000000   
31                     Niguarda           0.000000            0.000000   
32                       Ortica           0.000000            0.000000   
33                 Ponte Seveso           0.000000            0.000000   
34              Porta Garibaldi           0.022222            0.000000   
35               Porta Monforte           0.000000            0.000000   
36                  Porta Nuova           0.000000            0.000000   
37               Porta Sempione           0.000000            0.000000   
38             Porta Tenaglia\n           0.020408            0.000000   
39                Porta Venezia           0.000000            0.000000   
40                  Porta Volta           0.000000            0.000000   
41                     Portello           0.000000            0.000000   
42              Prato Centenaro           0.000000            0.000000   
43                     Precotto           0.000000            0.000000   
44                          QT8           0.000000            0.000000   
45             Quartiere Feltre           0.000000            0.000000   
46               Quarto Oggiaro           0.000000            0.000000   
47                      Roserio           0.000000            0.000000   
48                    Rottole\n           0.000000            0.000000   
49                 San Leonardo           0.000000            0.000000   
50                    Segnano\n           0.000000            0.000000   
51            Stazione Centrale           0.000000            0.000000   
52                       Trenno           0.000000            0.000000   
53                     

In [40]:
milan_grouped.shape

(58, 147)

Print each neighborhood along with the top 5 most common venues

In [41]:
num_top_venues = 10

for hood in milan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = milan_grouped[milan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adriano----
                 venue  freq
0          Supermarket  0.17
1  Sporting Goods Shop  0.17
2                 Park  0.17
3    Electronics Store  0.17
4          Pizza Place  0.17
5               Hostel  0.17
6      Photography Lab  0.00
7          Pastry Shop  0.00
8     Pedestrian Plaza  0.00
9             Pet Café  0.00


----Affori----
                   venue  freq
0               Bus Stop  0.11
1     Italian Restaurant  0.11
2  General Entertainment  0.11
3   Gym / Fitness Center  0.11
4      Health Food Store  0.11
5                 Hostel  0.11
6           Cocktail Bar  0.11
7                   Café  0.11
8                   Park  0.11
9            Pizza Place  0.00


----Bicocca----
               venue  freq
0               Café  0.26
1              Plaza  0.13
2       Tram Station  0.09
3  Trattoria/Osteria  0.04
4      Historic Site  0.04
5            Theater  0.04
6                Pub  0.04
7          Cafeteria  0.04
8         Piadineria  0.04
9   Kebab Restauran

In [42]:
milan_wine = milan_grouped [['Neighborhood','Wine Bar','Wine Shop']]
milan_wine

Neighborhood  Wine Bar  Wine Shop
0                       Adriano  0.000000   0.000000
1                        Affori  0.000000   0.000000
2                       Bicocca  0.000000   0.000000
3                        Bovisa  0.083333   0.000000
4                     Bovisasca  0.000000   0.000000
5                         Brera  0.020408   0.000000
6                      Bruzzano  0.000000   0.000000
7                    Ca' Granda  0.000000   0.000000
8                     Casoretto  0.000000   0.018868
9            Centro Direzionale  0.000000   0.000000
10                      Cimiano  0.000000   0.000000
11                  Città Studi  0.083333   0.000000
12                     Comasina  0.000000   0.000000
13           Conca del Naviglio  0.029412   0.000000
14                  Crescenzago  0.000000   0.000000
15                      Dergano  0.022222   0.022222
16                        Dosso  0.000000   0.000000
17                 Fulvio Testi  0.000000   0.000000
18                        Gorla  0.000000   0.000000
19                        Greco  0.000000   0.000000
20                    Guastalla  0.000000   0.000000
21                        Isola  0.083333   0.000000
22                   La Fontana  0.000000   0.000000
23                     Lambrate  0.000000   0.018868
24                   Lampugnano  0.000000   0.000000
25                       Loreto  0.000000   0.018868
26                   Maggiolina  0.000000   0.000000
27                     Mandello  0.029412   0.000000
28                    Mirabello  0.020408   0.000000
29                   Montalbino  0.000000   0.000000
30                      Musocco  0.000000   0.000000
31                     Niguarda  0.000000   0.000000
32                       Ortica  0.000000   0.000000
33                 Ponte Seveso  0.000000   0.000000
34              Porta Garibaldi  0.022222   0.022222
35               Porta Monforte  0.000000   0.000000
36                  Porta Nuova  0.000000   0.000000
37               Porta Sempione  0.000000   0.000000
38             Porta Tenaglia\n  0.020408   0.000000
39                Porta Venezia  0.026316   0.000000
40                  Porta Volta  0.000000   0.000000
41                     Portello  0.000000   0.000000
42              Prato Centenaro  0.000000   0.000000
43                     Precotto  0.000000   0.000000
44                          QT8  0.000000   0.000000
45             Quartiere Feltre  0.000000   0.000000
46               Quarto Oggiaro  0.000000   0.000000
47                      Roserio  0.000000   0.000000
48                    Rottole\n  0.000000   0.000000
49                 San Leonardo  0.000000   0.000000
50                    Segnano\n  0.000000   0.000000
51            Stazione Centrale  0.000000   0.000000
52                       Trenno  0.000000   0.000000
53                        Turro  0.000000   0.000000
54                     Varesina  0.000000   0.000000
55                       Vialba  0.000000   0.000000
56  Villaggio dei Giornalisti\n  0.000000   0.016667
57               Villapizzone\n  0.000000   0.000000

In [43]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [44]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = milan_grouped['Neighborhood']

for ind in np.arange(milan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(milan_grouped.iloc[ind, :], num_top_venues)
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0      Adriano   Sporting Goods Shop           Supermarket   
1       Affori  Gym / Fitness Center                Hostel   
2      Bicocca                  Café                 Plaza   
3       Bovisa                   Bar    Italian Restaurant   
4    Bovisasca          Tram Station           Event Space   

    3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0             Pizza Place                Hostel      Electronics Store   
1                    Café                  Park  General Entertainment   
2            Tram Station                   Pub                Theater   
3             Pizza Place              Wine Bar                    Pub   
4  Furniture / Home Store   Fried Chicken Joint               Fountain   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                  Park           Flower Shop   Fried Chicken Joint   
1          Cocktail Bar     Health Food Store              Bus Stop   
2            Restaurant        Sandwich Place             Cafeteria   
3                  Café         Train Station                  Park   
4            Food Court     Food & Drink Shop           Flower Shop   

  9th Most Common Venue 10th Most Common Venue  
0              Fountain             Food Court  
1    Italian Restaurant   Ethiopian Restaurant  
2            Piadineria            Supermarket  
3          Cocktail Bar            Event Space  
4  Fast Food Restaurant          Women's Store

In [45]:
neighborhoods_venues_sorted.rename(columns={'Neighborhood': 'Neighbourhoods'}, inplace=True)
neighborhoods_venues_sorted.head()

Neighbourhoods 1st Most Common Venue 2nd Most Common Venue  \
0        Adriano   Sporting Goods Shop           Supermarket   
1         Affori  Gym / Fitness Center                Hostel   
2        Bicocca                  Café                 Plaza   
3         Bovisa                   Bar    Italian Restaurant   
4      Bovisasca          Tram Station           Event Space   

    3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0             Pizza Place                Hostel      Electronics Store   
1                    Café                  Park  General Entertainment   
2            Tram Station                   Pub                Theater   
3             Pizza Place              Wine Bar                    Pub   
4  Furniture / Home Store   Fried Chicken Joint               Fountain   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                  Park           Flower Shop   Fried Chicken Joint   
1          Cocktail Bar     Health Food Store              Bus Stop   
2            Restaurant        Sandwich Place             Cafeteria   
3                  Café         Train Station                  Park   
4            Food Court     Food & Drink Shop           Flower Shop   

  9th Most Common Venue 10th Most Common Venue  
0              Fountain             Food Court  
1    Italian Restaurant   Ethiopian Restaurant  
2            Piadineria            Supermarket  
3          Cocktail Bar            Event Space  
4  Fast Food Restaurant          Women's Store

Cluster Neighborhoods

In [46]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [47]:
# set number of clusters
kclusters = 5

milan_grouped_clustering = milan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(milan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 0, 0, 0, 3, 0, 0, 0, 0, 1], dtype=int32)

In [51]:
# add clustering labels

neighborhoods_venues_sorted.insert(0, 'Cluster Labels2', kmeans.labels_)

milan_merged = Combined_trial

# mergem milan_grouped with milan_data to add latitude/longitude for each neighborhood
milan_merged = milan_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhoods'), on='Neighbourhoods')

milan_merged.head() 

Neighbourhoods  Postal Code   Latitude  Longitude  Cluster Labels2  \
0               Brera        20121  45.472955   9.187099                0   
1  Conca del Naviglio        20123  45.463065   9.177542                0   
2           Guastalla        20151  45.494555   9.113667                1   
3      Porta Sempione        20145  45.473887   9.161670                0   
4    Porta Tenaglia\n        20121  45.472955   9.187099                0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0    Italian Restaurant                 Hotel        Ice Cream Shop   
1    Italian Restaurant                  Café        Ice Cream Shop   
2  Fast Food Restaurant                  Café           Pizza Place   
3                 Plaza                  Park          Cocktail Bar   
4    Italian Restaurant                 Hotel        Ice Cream Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0               Theater      Sushi Restaurant   Japanese Restaurant   
1          Dessert Shop           Pizza Place        Sandwich Place   
2           Supermarket                  Park        Clothing Store   
3                  Café                   Bar           Supermarket   
4               Theater      Sushi Restaurant   Japanese Restaurant   

  7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
0        Sandwich Place  College Arts Building          Cupcake Shop   
1                 Plaza                  Hotel                Hostel   
2             Bookstore                    Gym         Metro Station   
3      Sushi Restaurant     Italian Restaurant          Tennis Court   
4        Sandwich Place  College Arts Building          Cupcake Shop   

  10th Most Common Venue  
0     Salon / Barbershop  
1              Gift Shop  
2                   Pool  
3         Cosmetics Shop  
4     Salon / Barbershop

In [52]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [53]:
milan_merged['Cluster Labels2'].isna().sum()

0

In [54]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(milan_merged['Latitude'],milan_merged['Longitude'], milan_merged['Neighbourhoods'], milan_merged['Cluster Labels2']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color = rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters


let's understand more on the nature of such clusters

In [71]:
milan_merged.loc[milan_merged['Cluster Labels2'] == 0, milan_merged.columns[[1] + list(range(5, milan_merged.shape[1]))]]


Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0         20121    Italian Restaurant                 Hotel   
1         20123    Italian Restaurant                  Café   
3         20145                 Plaza                  Park   
4         20121    Italian Restaurant                 Hotel   
7         20126                  Café                 Plaza   
8         20125            Playground            Kids Store   
9         20131    Italian Restaurant    Chinese Restaurant   
10        20125            Playground            Kids Store   
11        20123    Italian Restaurant                  Café   
12        20121    Italian Restaurant                 Hotel   
13        20125            Playground            Kids Store   
14        20125            Playground            Kids Store   
16        20125            Playground            Kids Store   
17        20127    Italian Restaurant           Pizza Place   
18        20129    Italian Restaurant                  Café   
19        20131    Italian Restaurant    Chinese Restaurant   
21        20077                   Bar    Italian Restaurant   
23        20131    Italian Restaurant    Chinese Restaurant   
25        20122        Clothing Store    Italian Restaurant   
26        20143                  Café    Italian Restaurant   
41        20161  Gym / Fitness Center                Hostel   
42        20126                  Café                 Plaza   
43        20077                   Bar    Italian Restaurant   
45        20161  Gym / Fitness Center                Hostel   
46        20162                   Pub    Italian Restaurant   
48        20161  Gym / Fitness Center                Hostel   
49        20154    Chinese Restaurant    Italian Restaurant   
50        20124    Italian Restaurant                 Hotel   
51        20077                   Bar    Italian Restaurant   
52        20122        Clothing Store    Italian Restaurant   
53        20159           Pizza Place          Tram Station   
54        20162                   Pub    Italian Restaurant   
55        20154    Chinese Restaurant    Italian Restaurant   
56        20125            Playground            Kids Store   
57        20125            Playground            Kids Store   
58        20060           Music Venue                   Bar   

   3rd Most Common Venue   4th Most Common Venue  5th Most Common Venue  \
0         Ice Cream Shop                 Theater       Sushi Restaurant   
1         Ice Cream Shop            Dessert Shop            Pizza Place   
3           Cocktail Bar                    Café                    Bar   
4         Ice Cream Shop                 Theater       Sushi Restaurant   
7           Tram Station                     Pub                Theater   
8            Pizza Place      Chinese Restaurant             Steakhouse   
9            Pizza Place          Ice Cream Shop    Japanese Restaurant   
10           Pizza Place      Chinese Restaurant             Steakhouse   
11        Ice Cream Shop            Dessert Shop            Pizza Place   
12        Ice Cream Shop                 Theater       Sushi Restaurant   
13           Pizza Place      Chinese Restaurant             Steakhouse   
14           Pizza Place      Chinese Restaurant             Steakhouse   
16           Pizza Place      Chinese Restaurant             Steakhouse   
17     Indian Restaurant          Sandwich Place            Music Venue   
18      Sushi Restaurant             Pizza Place           Cocktail Bar   
19           Pizza Place          Ice Cream Shop    Japanese Restaurant   
21           Pizza Place                Wine Bar                    Pub   
23           Pizza Place          Ice Cream Shop    Japanese Restaurant   
25                 Plaza  Furniture / Home Store    Sporting Goods Shop   
26           Supermarket          Ice Cream Shop            Pizza Place   
41                  Café                    Park  General Entertainment   
42          Tram Stati

In [56]:
milan_merged.loc[milan_merged['Cluster Labels2'] == 1, milan_merged.columns[[1] + list(range(5, milan_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
2         20151  Fast Food Restaurant                  Café   
5         20128   Sporting Goods Shop           Supermarket   
15        20128   Sporting Goods Shop           Supermarket   
29        20151  Fast Food Restaurant                  Café   
30        20151  Fast Food Restaurant                  Café   
31        20151  Fast Food Restaurant                  Café   
32        20151  Fast Food Restaurant                  Café   
33        20151  Fast Food Restaurant                  Café   
34        20151  Fast Food Restaurant                  Café   
35        20151  Fast Food Restaurant                  Café   
36        20151  Fast Food Restaurant                  Café   
37        20151  Fast Food Restaurant                  Café   
38        20151  Fast Food Restaurant                  Café   
39        20151  Fast Food Restaurant                  Café   
40        20151  Fast Food Restaurant                  Café   
47        20090                 Beach          Cocktail Bar   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2            Pizza Place           Supermarket                  Park   
5            Pizza Place                Hostel     Electronics Store   
15           Pizza Place                Hostel     Electronics Store   
29           Pizza Place           Supermarket                  Park   
30           Pizza Place           Supermarket                  Park   
31           Pizza Place           Supermarket                  Park   
32           Pizza Place           Supermarket                  Park   
33           Pizza Place           Supermarket                  Park   
34           Pizza Place           Supermarket                  Park   
35           Pizza Place           Supermarket                  Park   
36           Pizza Place           Supermarket                  Park   
37           Pizza Place           Supermarket                  Park   
38           Pizza Place           Supermarket                  Park   
39           Pizza Place           Supermarket                  Park   
40           Pizza Place           Supermarket                  Park   
47           Supermarket                  Café            Shoe Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2         Clothing Store             Bookstore                   Gym   
5                   Park           Flower Shop   Fried Chicken Joint   
15                  Park           Flower Shop   Fried Chicken Joint   
29        Clothing Store             Bookstore                   Gym   
30        Clothing Store             Bookstore                   Gym   
31        Clothing Store             Bookstore                   Gym   
32        Clothing Store             Bookstore                   Gym   
33        Clothing Store             Bookstore                   Gym   
34        Clothing Store             Bookstore                   Gym   
35        Clothing Store             Bookstore                   Gym   
36        Clothing Store             Bookstore                   Gym   
37        Clothing Store             Bookstore                   Gym   
38        Clothing Store             Bookstore                   Gym   
39        Clothing Store             Bookstore                   Gym   
40        Clothing Store             Bookstore                   Gym   
47  Gym / Fitness Center                   Spa           Pizza Place   

   9th Most Common Venue 10th Most Common Venue  
2          Metro Station                   Pool  
5               Fountain             Food Court  
15              Fountain             Food Court  
29         Metro Station                   Pool  
30         Metro Station                   Pool  
31         Metro Station                   Pool  
32         Metro Station                   Pool  
33         Metro Station                   Pool  
34         Metro Station        

In [57]:
milan_merged.loc[milan_merged['Cluster Labels2'] == 2, milan_merged.columns[[1] + list(range(5, milan_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
6         20134      Sushi Restaurant                  Park   
24        20134      Sushi Restaurant                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6                    Gym             Nightclub         Women's Store   
24                   Gym             Nightclub         Women's Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6            Flower Shop   Fried Chicken Joint              Fountain   
24           Flower Shop   Fried Chicken Joint              Fountain   

   9th Most Common Venue 10th Most Common Venue  
6             Food Court      Food & Drink Shop  
24            Food Court      Food & Drink Shop

In [58]:
milan_merged.loc[milan_merged['Cluster Labels2'] == 3, milan_merged.columns[[1] + list(range(5, milan_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
44        20157          Tram Station           Event Space   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
44  Furniture / Home Store   Fried Chicken Joint              Fountain   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
44            Food Court     Food & Drink Shop           Flower Shop   

   9th Most Common Venue 10th Most Common Venue  
44  Fast Food Restaurant          Women's Store

In [59]:
milan_merged.loc[milan_merged['Cluster Labels2'] == 4, milan_merged.columns[[1] + list(range(5, milan_merged.shape[1]))]]

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
20        20132           Pizza Place                  Café   
22        20132           Pizza Place                  Café   
27        20132           Pizza Place                  Café   
28        20132           Pizza Place                  Café   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
20                 Hotel    Italian Restaurant           Event Space   
22                 Hotel    Italian Restaurant           Event Space   
27                 Hotel    Italian Restaurant           Event Space   
28                 Hotel    Italian Restaurant           Event Space   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
20              Fountain            Food Court     Food & Drink Shop   
22              Fountain            Food Court     Food & Drink Shop   
27              Fountain            Food Court     Food & Drink Shop   
28              Fountain            Food Court     Food & Drink Shop   

   9th Most Common Venue 10th Most Common Venue  
20           Flower Shop   Fast Food Restaurant  
22           Flower Shop   Fast Food Restaurant  
27           Flower Shop   Fast Food Restaurant  
28           Flower Shop   Fast Food Restaurant

In [60]:
milan_wine2= milan_wine

In [61]:
milan_wine2

Neighborhood  Wine Bar  Wine Shop
0                       Adriano  0.000000   0.000000
1                        Affori  0.000000   0.000000
2                       Bicocca  0.000000   0.000000
3                        Bovisa  0.083333   0.000000
4                     Bovisasca  0.000000   0.000000
5                         Brera  0.020408   0.000000
6                      Bruzzano  0.000000   0.000000
7                    Ca' Granda  0.000000   0.000000
8                     Casoretto  0.000000   0.018868
9            Centro Direzionale  0.000000   0.000000
10                      Cimiano  0.000000   0.000000
11                  Città Studi  0.083333   0.000000
12                     Comasina  0.000000   0.000000
13           Conca del Naviglio  0.029412   0.000000
14                  Crescenzago  0.000000   0.000000
15                      Dergano  0.022222   0.022222
16                        Dosso  0.000000   0.000000
17                 Fulvio Testi  0.000000   0.000000
18                        Gorla  0.000000   0.000000
19                        Greco  0.000000   0.000000
20                    Guastalla  0.000000   0.000000
21                        Isola  0.083333   0.000000
22                   La Fontana  0.000000   0.000000
23                     Lambrate  0.000000   0.018868
24                   Lampugnano  0.000000   0.000000
25                       Loreto  0.000000   0.018868
26                   Maggiolina  0.000000   0.000000
27                     Mandello  0.029412   0.000000
28                    Mirabello  0.020408   0.000000
29                   Montalbino  0.000000   0.000000
30                      Musocco  0.000000   0.000000
31                     Niguarda  0.000000   0.000000
32                       Ortica  0.000000   0.000000
33                 Ponte Seveso  0.000000   0.000000
34              Porta Garibaldi  0.022222   0.022222
35               Porta Monforte  0.000000   0.000000
36                  Porta Nuova  0.000000   0.000000
37               Porta Sempione  0.000000   0.000000
38             Porta Tenaglia\n  0.020408   0.000000
39                Porta Venezia  0.026316   0.000000
40                  Porta Volta  0.000000   0.000000
41                     Portello  0.000000   0.000000
42              Prato Centenaro  0.000000   0.000000
43                     Precotto  0.000000   0.000000
44                          QT8  0.000000   0.000000
45             Quartiere Feltre  0.000000   0.000000
46               Quarto Oggiaro  0.000000   0.000000
47                      Roserio  0.000000   0.000000
48                    Rottole\n  0.000000   0.000000
49                 San Leonardo  0.000000   0.000000
50                    Segnano\n  0.000000   0.000000
51            Stazione Centrale  0.000000   0.000000
52                       Trenno  0.000000   0.000000
53                        Turro  0.000000   0.000000
54                     Varesina  0.000000   0.000000
55                       Vialba  0.000000   0.000000
56  Villaggio dei Giornalisti\n  0.000000   0.016667
57               Villapizzone\n  0.000000   0.000000

Combine Milan_data with Wine data to conclude

In [62]:
Combined_trial.rename(columns={'Neighbourhoods': 'Neighborhood'}, inplace=True)
Combined_trial

Neighborhood  Postal Code   Latitude  Longitude
0                         Brera        20121  45.472955   9.187099
1            Conca del Naviglio        20123  45.463065   9.177542
2                     Guastalla        20151  45.494555   9.113667
3                Porta Sempione        20145  45.473887   9.161670
4              Porta Tenaglia\n        20121  45.472955   9.187099
5                       Adriano        20128  45.513235   9.239450
6                   Crescenzago        20134  45.482773   9.254800
7                         Gorla        20126  45.515929   9.215605
8                         Greco        20125  45.499710   9.208823
9                        Loreto        20131  45.487090   9.227062
10                   Maggiolina        20125  45.499710   9.208823
11                     Mandello        20123  45.463065   9.177542
12                    Mirabello        20121  45.472955   9.187099
13                 Ponte Seveso        20125  45.499710   9.208823
14                  Porta Nuova        20125  45.499710   9.208823
15                     Precotto        20128  45.513235   9.239450
16            Stazione Centrale        20125  45.499710   9.208823
17                        Turro        20127  45.497700   9.226106
18  Villaggio dei Giornalisti\n        20129  45.472203   9.213030
19                    Casoretto        20131  45.487090   9.227062
20                      Cimiano        20132  45.503230   9.246742
21                  Città Studi        20077  45.355305   9.320114
22                        Dosso        20132  45.503230   9.246742
23                     Lambrate        20131  45.487090   9.227062
24                       Ortica        20134  45.482773   9.254800
25               Porta Monforte        20122  45.462565   9.198647
26                Porta Venezia        20143  45.444575   9.158881
27             Quartiere Feltre        20132  45.503230   9.246742
28                    Rottole\n        20132  45.503230   9.246742
29                   Lampugnano        20151  45.494555   9.113667
30                      Musocco        20151  45.494555   9.113667
31                  Porta Volta        20151  45.494555   9.113667
32                     Portello        20151  45.494555   9.113667
33               Quarto Oggiaro        20151  45.494555   9.113667
34                          QT8        20151  45.494555   9.113667
35                      Roserio        20151  45.494555   9.113667
36                 San Leonardo        20151  45.494555   9.113667
37                       Trenno        20151  45.494555   9.113667
38                     Varesina        20151  45.494555   9.113667
39                       Vialba        20151  45.494555   9.113667
40               Villapizzone\n        20151  45.494555   9.113667
41                       Affori        20161  45.520290   9.166989
42                      Bicocca        20126  45.515929   9.215605
43                       Bovisa        20077  45.355305   9.320114
44                    Bovisasca        20157  45.516985   9.125239
45                     Bruzzano        20161  45.520290   9.166989
46                   Ca' Granda        20162  45.515339   9.197820
47           Centro Direzionale        20090  45.423010   9.055915
48                     Comasina        20161  45.520290   9.166989
49                      Dergano        20154  45.483920   9.175276
50                 Fulvio Testi        20124  45.485165   9.201855
51                        Isola        20077  45.355305   9.320114
52                   La Fontana        20122  45.462565   9.198647
53                   Montalbino        20159  45.496732   9.189580
54                     Niguarda        20162  45.515339   9.197820
55              Porta Garibaldi        20154  45.483920   9.175276
56                  Porta Nuova        20125  45.499710   9.208823
57              Prato Centenaro        20125  45.499710   9.208823
58                    Segnano\n        20060  45.523685   9.449335

In [63]:
milan_overall = milan_wine.join(Combined_trial.set_index('Neighborhood'), on='Neighborhood')
milan_overall

Neighborhood  Wine Bar  Wine Shop  Postal Code   Latitude  \
0                       Adriano  0.000000   0.000000        20128  45.513235   
1                        Affori  0.000000   0.000000        20161  45.520290   
2                       Bicocca  0.000000   0.000000        20126  45.515929   
3                        Bovisa  0.083333   0.000000        20077  45.355305   
4                     Bovisasca  0.000000   0.000000        20157  45.516985   
5                         Brera  0.020408   0.000000        20121  45.472955   
6                      Bruzzano  0.000000   0.000000        20161  45.520290   
7                    Ca' Granda  0.000000   0.000000        20162  45.515339   
8                     Casoretto  0.000000   0.018868        20131  45.487090   
9            Centro Direzionale  0.000000   0.000000        20090  45.423010   
10                      Cimiano  0.000000   0.000000        20132  45.503230   
11                  Città Studi  0.083333   0.000000        20077  45.355305   
12                     Comasina  0.000000   0.000000        20161  45.520290   
13           Conca del Naviglio  0.029412   0.000000        20123  45.463065   
14                  Crescenzago  0.000000   0.000000        20134  45.482773   
15                      Dergano  0.022222   0.022222        20154  45.483920   
16                        Dosso  0.000000   0.000000        20132  45.503230   
17                 Fulvio Testi  0.000000   0.000000        20124  45.485165   
18                        Gorla  0.000000   0.000000        20126  45.515929   
19                        Greco  0.000000   0.000000        20125  45.499710   
20                    Guastalla  0.000000   0.000000        20151  45.494555   
21                        Isola  0.083333   0.000000        20077  45.355305   
22                   La Fontana  0.000000   0.000000        20122  45.462565   
23                     Lambrate  0.000000   0.018868        20131  45.487090   
24                   Lampugnano  0.000000   0.000000        20151  45.494555   
25                       Loreto  0.000000   0.018868        20131  45.487090   
26                   Maggiolina  0.000000   0.000000        20125  45.499710   
27                     Mandello  0.029412   0.000000        20123  45.463065   
28                    Mirabello  0.020408   0.000000        20121  45.472955   
29                   Montalbino  0.000000   0.000000        20159  45.496732   
30                      Musocco  0.000000   0.000000        20151  45.494555   
31                     Niguarda  0.000000   0.000000        20162  45.515339   
32                       Ortica  0.000000   0.000000        20134  45.482773   
33                 Ponte Seveso  0.000000   0.000000        20125  45.499710   
34              Porta Garibaldi  0.022222   0.022222        20154  45.483920   
35               Porta Monforte  0.000000   0.000000        20122  45.462565   
36                  Porta Nuova  0.000000   0.000000        20125  45.499710   
36                  Porta Nuova  0.000000   0.000000        20125  45.499710   
37               Porta Sempione  0.000000   0.000000        20145  45.473887   
38             Porta Tenaglia\n  0.020408   0.000000        20121  45.472955   
39                Porta Venezia  0.026316   0.000000        20143  45.444575   
40                  Porta Volta  0.000000   0.000000        20151  45.494555   
41                     Portello  0.000000   0.000000        20151  45.494555   
42              Prato Centenaro  0.000000   0.000000        20125  45.499710   
43                     Precotto  0.000000   0.000000        20128  45.513235   
44                          QT8  0.000000   0.000000        20151  45.494555   
45             Quartiere Feltre  0.000000   0.000000        20132  45.503230   
46               Quarto Oggiaro  0.000000   0.000000        20151  45.494555   
47                      Roserio  0.000000   0.000000        20151  45.494555   
48                 

In [65]:
milan_merged_2 = milan_merged.join(x1.set_index('Neighbourhoods'), on='Neighbourhoods')

milan_merged_2


Neighbourhoods  Postal Code   Latitude  Longitude  \
0                         Brera        20121  45.472955   9.187099   
1            Conca del Naviglio        20123  45.463065   9.177542   
2                     Guastalla        20151  45.494555   9.113667   
3                Porta Sempione        20145  45.473887   9.161670   
4              Porta Tenaglia\n        20121  45.472955   9.187099   
5                       Adriano        20128  45.513235   9.239450   
6                   Crescenzago        20134  45.482773   9.254800   
7                         Gorla        20126  45.515929   9.215605   
8                         Greco        20125  45.499710   9.208823   
9                        Loreto        20131  45.487090   9.227062   
10                   Maggiolina        20125  45.499710   9.208823   
11                     Mandello        20123  45.463065   9.177542   
12                    Mirabello        20121  45.472955   9.187099   
13                 Ponte Seveso        20125  45.499710   9.208823   
14                  Porta Nuova        20125  45.499710   9.208823   
15                     Precotto        20128  45.513235   9.239450   
16            Stazione Centrale        20125  45.499710   9.208823   
17                        Turro        20127  45.497700   9.226106   
18  Villaggio dei Giornalisti\n        20129  45.472203   9.213030   
19                    Casoretto        20131  45.487090   9.227062   
20                      Cimiano        20132  45.503230   9.246742   
21                  Città Studi        20077  45.355305   9.320114   
22                        Dosso        20132  45.503230   9.246742   
23                     Lambrate        20131  45.487090   9.227062   
24                       Ortica        20134  45.482773   9.254800   
25               Porta Monforte        20122  45.462565   9.198647   
26                Porta Venezia        20143  45.444575   9.158881   
27             Quartiere Feltre        20132  45.503230   9.246742   
28                    Rottole\n        20132  45.503230   9.246742   
29                   Lampugnano        20151  45.494555   9.113667   
30                      Musocco        20151  45.494555   9.113667   
31                  Porta Volta        20151  45.494555   9.113667   
32                     Portello        20151  45.494555   9.113667   
33               Quarto Oggiaro        20151  45.494555   9.113667   
34                          QT8        20151  45.494555   9.113667   
35                      Roserio        20151  45.494555   9.113667   
36                 San Leonardo        20151  45.494555   9.113667   
37                       Trenno        20151  45.494555   9.113667   
38                     Varesina        20151  45.494555   9.113667   
39                       Vialba        20151  45.494555   9.113667   
40               Villapizzone\n        20151  45.494555   9.113667   
41                       Affori        20161  45.520290   9.166989   
42                      Bicocca        20126  45.515929   9.215605   
43                       Bovisa        20077  45.355305   9.320114   
44                    Bovisasca        20157  45.516985   9.125239   
45                     Bruzzano        20161  45.520290   9.166989   
46                   Ca' Granda        20162  45.515339   9.197820   
47           Centro Direzionale        20090  45.423010   9.055915   
48                     Comasina        20161  45.520290   9.166989   
49                      Dergano        20154  45.483920   9.175276   
50                 Fulvio Testi        20124  45.485165   9.201855   
51                        Isola        20077  45.355305   9.320114   
52                   La Fontana        20122  45.462565   9.198647   
53                   Montalbino        20159  45.496732   9.189580   
54                     Niguarda        20162  45.515339   9.197820   
55              Porta Garibaldi        20154  45.483920   9.175276   
56                  Porta N

In [68]:
milan_merged_2 = milan_merged_2.sort_values('Density population_inhabitans_km2', ascending=False)
milan_merged_2

Neighbourhoods  Postal Code   Latitude  Longitude  \
41                       Affori        20161  45.520290   9.166989   
5                       Adriano        20128  45.513235   9.239450   
0                         Brera        20121  45.472955   9.187099   
19                    Casoretto        20131  45.487090   9.227062   
1            Conca del Naviglio        20123  45.463065   9.177542   
2                     Guastalla        20151  45.494555   9.113667   
3                Porta Sempione        20145  45.473887   9.161670   
4              Porta Tenaglia\n        20121  45.472955   9.187099   
6                   Crescenzago        20134  45.482773   9.254800   
7                         Gorla        20126  45.515929   9.215605   
8                         Greco        20125  45.499710   9.208823   
9                        Loreto        20131  45.487090   9.227062   
10                   Maggiolina        20125  45.499710   9.208823   
11                     Mandello        20123  45.463065   9.177542   
12                    Mirabello        20121  45.472955   9.187099   
13                 Ponte Seveso        20125  45.499710   9.208823   
14                  Porta Nuova        20125  45.499710   9.208823   
15                     Precotto        20128  45.513235   9.239450   
16            Stazione Centrale        20125  45.499710   9.208823   
17                        Turro        20127  45.497700   9.226106   
18  Villaggio dei Giornalisti\n        20129  45.472203   9.213030   
20                      Cimiano        20132  45.503230   9.246742   
21                  Città Studi        20077  45.355305   9.320114   
22                        Dosso        20132  45.503230   9.246742   
23                     Lambrate        20131  45.487090   9.227062   
24                       Ortica        20134  45.482773   9.254800   
25               Porta Monforte        20122  45.462565   9.198647   
26                Porta Venezia        20143  45.444575   9.158881   
27             Quartiere Feltre        20132  45.503230   9.246742   
28                    Rottole\n        20132  45.503230   9.246742   
29                   Lampugnano        20151  45.494555   9.113667   
30                      Musocco        20151  45.494555   9.113667   
31                  Porta Volta        20151  45.494555   9.113667   
32                     Portello        20151  45.494555   9.113667   
33               Quarto Oggiaro        20151  45.494555   9.113667   
34                          QT8        20151  45.494555   9.113667   
35                      Roserio        20151  45.494555   9.113667   
36                 San Leonardo        20151  45.494555   9.113667   
37                       Trenno        20151  45.494555   9.113667   
38                     Varesina        20151  45.494555   9.113667   
39                       Vialba        20151  45.494555   9.113667   
40               Villapizzone\n        20151  45.494555   9.113667   
42                      Bicocca        20126  45.515929   9.215605   
43                       Bovisa        20077  45.355305   9.320114   
44                    Bovisasca        20157  45.516985   9.125239   
45                     Bruzzano        20161  45.520290   9.166989   
46                   Ca' Granda        20162  45.515339   9.197820   
47           Centro Direzionale        20090  45.423010   9.055915   
48                     Comasina        20161  45.520290   9.166989   
49                      Dergano        20154  45.483920   9.175276   
50                 Fulvio Testi        20124  45.485165   9.201855   
51                        Isola        20077  45.355305   9.320114   
52                   La Fontana        20122  45.462565   9.198647   
53                   Montalbino        20159  45.496732   9.189580   
54                     Niguarda        20162  45.515339   9.197820   
55              Porta Garibaldi        20154  45.483920   9.175276   
56                  Porta N

In [72]:
milan_merged_2 = milan_merged_2.sort_values('Cluster Labels2', ascending=False)
milan_merged_2

Neighbourhoods  Postal Code   Latitude  Longitude  \
28                    Rottole\n        20132  45.503230   9.246742   
20                      Cimiano        20132  45.503230   9.246742   
27             Quartiere Feltre        20132  45.503230   9.246742   
22                        Dosso        20132  45.503230   9.246742   
44                    Bovisasca        20157  45.516985   9.125239   
6                   Crescenzago        20134  45.482773   9.254800   
24                       Ortica        20134  45.482773   9.254800   
2                     Guastalla        20151  45.494555   9.113667   
40               Villapizzone\n        20151  45.494555   9.113667   
34                          QT8        20151  45.494555   9.113667   
33               Quarto Oggiaro        20151  45.494555   9.113667   
32                     Portello        20151  45.494555   9.113667   
31                  Porta Volta        20151  45.494555   9.113667   
30                      Musocco        20151  45.494555   9.113667   
29                   Lampugnano        20151  45.494555   9.113667   
5                       Adriano        20128  45.513235   9.239450   
37                       Trenno        20151  45.494555   9.113667   
38                     Varesina        20151  45.494555   9.113667   
39                       Vialba        20151  45.494555   9.113667   
35                      Roserio        20151  45.494555   9.113667   
36                 San Leonardo        20151  45.494555   9.113667   
15                     Precotto        20128  45.513235   9.239450   
47           Centro Direzionale        20090  45.423010   9.055915   
42                      Bicocca        20126  45.515929   9.215605   
48                     Comasina        20161  45.520290   9.166989   
53                   Montalbino        20159  45.496732   9.189580   
52                   La Fontana        20122  45.462565   9.198647   
51                        Isola        20077  45.355305   9.320114   
50                 Fulvio Testi        20124  45.485165   9.201855   
49                      Dergano        20154  45.483920   9.175276   
57              Prato Centenaro        20125  45.499710   9.208823   
54                     Niguarda        20162  45.515339   9.197820   
46                   Ca' Granda        20162  45.515339   9.197820   
45                     Bruzzano        20161  45.520290   9.166989   
56                  Porta Nuova        20125  45.499710   9.208823   
43                       Bovisa        20077  45.355305   9.320114   
55              Porta Garibaldi        20154  45.483920   9.175276   
41                       Affori        20161  45.520290   9.166989   
26                Porta Venezia        20143  45.444575   9.158881   
10                   Maggiolina        20125  45.499710   9.208823   
0                         Brera        20121  45.472955   9.187099   
19                    Casoretto        20131  45.487090   9.227062   
1            Conca del Naviglio        20123  45.463065   9.177542   
3                Porta Sempione        20145  45.473887   9.161670   
4              Porta Tenaglia\n        20121  45.472955   9.187099   
7                         Gorla        20126  45.515929   9.215605   
8                         Greco        20125  45.499710   9.208823   
9                        Loreto        20131  45.487090   9.227062   
11                     Mandello        20123  45.463065   9.177542   
25               Porta Monforte        20122  45.462565   9.198647   
12                    Mirabello        20121  45.472955   9.187099   
13                 Ponte Seveso        20125  45.499710   9.208823   
14                  Porta Nuova        20125  45.499710   9.208823   
16            Stazione Centrale        20125  45.499710   9.208823   
17                        Turro        20127  45.497700   9.226106   
18  Villaggio dei Giornalisti\n        20129  45.472203   9.213030   
21                  Città S

## Observations 

Cluster 0 has most food & entertainment businesses and many cocktail bars which suggest an intense nigh life. 
Cluster 1 has also similar businesses however clothing store, bookstore are also available. 
Wine shop conceived as a re-seller of wine product shall be treated differently from wine or cocktail bar and are recommended in area like those identified in Cluster 1. 

Brera is on the citycenter of Milan, clustered in group 0 as Affori with many more entertainement facilities and with high Density population/inhabitans/km2. In this area wine bar frequqncy is 0.020, however no wine shops as striclty conceived are available. 

## Conclusion 

 Adriano neighbourhood which is under Cluster 1 is a good candidate which has also high Density population/inhabitans/km2. It’s located in the north east of the city, out from city center which also suggest in principle affordable rental prices. Frequency of Wine shops and wine bars, according to Foursquare are null, therefore it’s definitely the recommended location to start up wine shop  business. 